tokio/src/io/interest.rs

```diff
@@ -116,12 +164,51 @@ impl Interest {
     /// assert!(BOTH.is_readable());
     /// assert!(BOTH.is_writable());
     pub const fn add(self, other: Interest) -> Interest {
-        Interest(self.0.add(other.0))
+        Self(self.0 | other.0)
     }
 
     // This function must be crate-private to avoid exposing a `mio` dependency.
-    pub(crate) const fn to_mio(self) -> mio::Interest {
-        self.0
+    pub(crate) fn to_mio(self) -> mio::Interest {
+        // mio does not allow and empty interest, so pick an arbitrary starting value
+        let mut mio = mio::Interest::READABLE;
+
+        if self.is_readable() {
+            mio |= mio::Interest::READABLE;
+        }
+
+        if self.is_writable() {
+            mio |= mio::Interest::WRITABLE;
+        }
+
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        if self.is_priority() {
+            mio |= mio::Interest::PRIORITY;
+        }
+
+        #[cfg(target_os = "freebsd")]
+        if self.is_aio() {
+            mio |= mio::Interest::AIO;
+        }
+
+        #[cfg(target_os = "freebsd")]
+        if self.is_lio() {
+            mio |= mio::Interest::LIO;
+        }
+
+        // Finally remove the starting value unless either:
+        //
+        // - the readable interest is specified
+        // - the error interest is specified. There is no error interest in mio, because error
+        //    events are always reported. But mio interests cannot be empty and one is needed
+        //    just for the registeration.
+        //
+        //    read readiness is filtered out in `Interest::mask` or `Ready::from_interest` if
+        //    needed
+        if !self.is_readable() && !self.is_error() {
+            mio.remove(mio::Interest::READABLE);
```

Comment:
This is incorrect.
```suggestion
            mio = mio.remove(mio::Interest::READABLE);
```

Reply:
good catch! `#[must_use]` on `remove` would be nice to prevent this sort of thing in the future. I'll make a `mio` PR for that

Reply:
I already made a PR.

Reply:
ah, I see it exists already, nice!